### PR TITLE
Add basic authentication support for fetching from Maven repositories

### PR DESCRIPTION
**Pull request by mrkane27**
Summary:

Ref #664.

Adds the ability to specify credentials when using Maven repositories.
These are passed using basic authentication.

One question would be whether to use `Authenticator` instead. The
worry there is that it's a global object, so I wasn't sure whether
I wanted to go that far. RFC.

Test Plan:

Have a `.buckconfig` file that looks like:

    ...
    [maven_repositories]
    central=https://repo1.maven.org/maven2
    foo=https://example.com/content/repositories/releases

    [credentials]
    foo_user=bar
    foo_pass=baz
    ...

Successfully ran `buck fetch` against a SonaType Nexus server configured to
use authentication.

### CODE REVIEW

File: src/com/facebook/buck/file/HttpDownloader.java [hunk 1]
```diff
@@ -57,6 +68,14 @@ public boolean fetch(BuckEventBus eventBus, URI uri, Path output) throws IOExcep
 
     try {
       HttpURLConnection connection = createConnection(uri);
+
+      if (authentication.isPresent()) {
+        PasswordAuthentication p = authentication.get();
+        String authStr = p.getUserName() + ":" + new String(p.getPassword());
+        String authEncoded = BaseEncoding.base64().encode(authStr.getBytes());
+        connection.addRequestProperty("Authorization", "Basic " + authEncoded);
```
Comment (Coneko) [inline src/com/facebook/buck/file/HttpDownloader.java]: Can you add a check for the connection being HTTPS? Sending passwords in clear text sounds pretty wrong.

**Comment by Coneko** [comment]
Can you add a test for this?

**Comment by mrkane27** [comment]
Will do! Is the overall approach reasonable?

File: src/com/facebook/buck/cli/DownloadConfig.java [hunk 2]
```diff
@@ -52,13 +53,13 @@ public DownloadConfig(BuckConfig delegate) {
     return delegate.getValue("download", "maven_repo");
   }
 
-  public ImmutableSet<String> getAllMavenRepos() {
-    ImmutableSortedSet.Builder<String> repos = ImmutableSortedSet.naturalOrder();
-    repos.addAll(delegate.getEntriesForSection("maven_repositories").values());
+  public ImmutableMap<String, String> getAllMavenRepos() {
+    ImmutableSortedMap.Builder<String, String> repos = ImmutableSortedMap.naturalOrder();
```
Comment (mrkane27) [inline src/com/facebook/buck/cli/DownloadConfig.java]: After this change, you could have multiple repos with the same URI but different aliases. Should I perhaps use a `BiMap` instead?

Comment (Coneko) [inline src/com/facebook/buck/cli/DownloadConfig.java]: Should be ok as is, I don't think it's an error to have that.

File: src/com/facebook/buck/file/HttpDownloader.java [hunk 3]
```diff
@@ -57,6 +68,19 @@ public boolean fetch(BuckEventBus eventBus, URI uri, Path output) throws IOExcep
 
     try {
       HttpURLConnection connection = createConnection(uri);
+
+      if (authentication.isPresent()) {
+        if ("https".equals(uri.getScheme())) {
```
Comment (Coneko) [inline src/com/facebook/buck/file/HttpDownloader.java]: Also check that the connections is an instance of https://docs.oracle.com/javase/7/docs/api/javax/net/ssl/HttpsURLConnection.html .

**Comment by Coneko** [comment]
Yeah I think the approach is fine, just fix that last thing and I'll import it.

**Comment by Coneko** [comment]
@facebook-github-bot shipit

**Comment by Coneko** [comment]
Bah, Travis failed apt-get so it didn't show test results, but this is what I get from the internal CI:

```
FAILURE com.facebook.buck.file.HttpDownloaderTest shouldAddAuthenticationHeader: 
  Expectation failure on verify:
    HttpURLConnection.addRequestProperty("Authorization", capture(Nothing captured yet)): expected: 1, actual: 0
java.lang.AssertionError: 
  Expectation failure on verify:
    HttpURLConnection.addRequestProperty("Authorization", capture(Nothing captured yet)): expected: 1, actual: 0
	at org.easymock.internal.MocksControl.verify(MocksControl.java:241)
	at org.easymock.EasyMock.verify(EasyMock.java:2100)
	at com.facebook.buck.file.HttpDownloaderTest.shouldAddAuthenticationHeader(HttpDownloaderTest.java:93)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.facebook.buck.testrunner.SameThreadFailOnTimeout$1.call(SameThreadFailOnTimeout.java:45)
	at com.facebook.buck.testrunner.SameThreadFailOnTimeout$1.call(SameThreadFailOnTimeout.java:41)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:744)
```

**Comment by mrkane27** [comment]
Sorry about that :blush:.

**Comment by Coneko** [comment]
@facebook-github-bot shipit

**Pull request closed without merging**